src/cluster_asm.c

```diff
@@ -365,6 +368,153 @@ void asmTaskFree(asmTask *task) {
+ * Format: "task_id:source_node:dest_node:operation:state:slot_ranges"
+ * Returns a new asmTask on success, NULL on failure. */
+asmTask *asmTaskDeserialize(sds data) {
+    int count, idx = 0;
+    asmTask *task = NULL;
+    if (!data) return NULL;
+
+    sds *parts = sdssplitlen(data, sdslen(data), ":", 1, &count);
+    if (count < 6) goto err;
+
+    /* Parse task ID */
+    if (sdslen(parts[idx]) != CLUSTER_NAMELEN) goto err;
+    task = asmTaskCreate(parts[idx]);
+    if (!task) goto err;
+    idx++;
+
+    /* Parse source node ID */
+    if (sdslen(parts[idx]) != CLUSTER_NAMELEN) goto err;
+    memcpy(task->source, parts[idx], CLUSTER_NAMELEN);
+    idx++;
+
+    /* Parse destination node ID */
+    if (sdslen(parts[idx]) != CLUSTER_NAMELEN) goto err;
+    memcpy(task->dest, parts[idx], CLUSTER_NAMELEN);
+    idx++;
+
+    /* Parse operation type */
+    if (!strcasecmp(parts[idx], "import")) {
+        task->operation = ASM_IMPORT;
+    } else if (!strcasecmp(parts[idx], "migrate")) {
+        task->operation = ASM_MIGRATE;
+    } else {
+        goto err;
+    }
+    idx++;
+
+    /* Parse state */
+    task->state = ASM_NONE; /* Default state */
+    for (int state = ASM_NONE; state <= ASM_RDBCHANNEL_TRANSFER; state++) {
+        if (!strcasecmp(parts[idx], asmTaskStateToString(state))) {
+            task->state = state;
+            break;
+        }
+    }
+    idx++;
+
+    /* Parse slot ranges */
+    task->slot_ranges = slotRangeArrayFromString(parts[idx]);
+    if (!task->slot_ranges) goto err;
+    idx++;
+
+    sdsfreesplitres(parts, count);
+    return task;
+
+err:
+    if (task) asmTaskFree(task);
+    sdsfreesplitres(parts, count);
+    return NULL;
+}
+
+/* Notify slaves about ASM task information to maintain consistency during slot migration.
+ * This function sends a CLUSTER SYNCSLOTS CONF MASTER-TASK command to all connected slaves
+ * with the serialized task information. */
+void asmNotifySlavesStateChange(struct asmTask *task) {
+    if (!clusterNodeIsMaster(getMyClusterNode()) || listLength(server.slaves) == 0) return;
+
+    /* Create command arguments for CLUSTER SYNCSLOTS CONF MASTER-TASK */
+    robj *argv[5];
+    argv[0] = createStringObject("CLUSTER", 7);
+    argv[1] = createStringObject("SYNCSLOTS", 9);
+    argv[2] = createStringObject("CONF", 4);
+    argv[3] = createStringObject("MASTER-TASK", 11);
+    argv[4] = createObject(OBJ_STRING, asmTaskSerialize(task));
+
+    /* Send the command to all slaves */
+    replicationFeedSlaves(server.slaves, -1, argv, 5);
+
+    /* Clean up command objects */
+    for (int i = 0; i < 5; i++) {
+        decrRefCount(argv[i]);
+    }
+}
+
+/* The replica may ask for full sync in the middle of a migration task, and has
+ * no idea about ASM start and end. So we need to notify the replica about the
+ * current task. This message may be duplicated, but in replica, we only process
+ * when the event is changed. */
+void asmNotifySlavesStateOnFullSync(void) {
```

Comment:
If replica restarts in the middle of a migration after being notified here, it may miss migration event when it wakes up right? We are not saving migration info to disk. When replica wakes up, there will be no migration event on its own but from master POV, migration might be still inprogress. Is that correct? Am I missing something?

Maybe we should propagate task info also when a replica connects to us?

Reply:
OK, let me do it